### PR TITLE
RSDK-6282 [part 2]: add PID values to the sensor controlled base config

### PR DESCRIPTION
**Pull request by martha-johnston**
second part of RSDK-6282 that adds PID values to the sensor controlled base config. Also removes `useControlLoop` bool as per [RSDK-5355](https://viam.atlassian.net/browse/RSDK-5355)

[RSDK-5355]: https://viam.atlassian.net/browse/RSDK-5355?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 1]
```diff
@@ -158,11 +171,47 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
...
-		sb.controlLoopConfig = sb.createControlLoopConfig()
-		if err := sb.setupControlLoops(); err != nil {
-			return err
+	if sb.velocities != nil && len(newConf.ControlParameters) != 0 {
+		// assign linear and angular PID correctly based on the given type
+		var linear, angular basePIDConfig
+		for _, c := range newConf.ControlParameters {
+			if c.Type == typeLinVel {
+				linear = c
+			} else if c.Type == typeAngVel {
```
Comment (susmitaSanyal) [inline components/base/sensorcontrolled/sensorcontrolled.go:180]: Since we are only comparing between typeLinVel and typeAngVel, I would recommend that you get rid of the else if check and assign c to angular in an else statement instead. 
```
if c.Type == typeLinVel {
	linear = c
} else {
       angular = c 
}
```

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 2]
```diff
@@ -158,11 +171,47 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
...
+				angular = c
+			}
 		}
+
+		sb.mu.Unlock()
```
Comment (susmitaSanyal) [inline components/base/sensorcontrolled/sensorcontrolled.go:185]: Why are we unlocking here?

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 3]
```diff
@@ -158,11 +171,47 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
...
+		// check if both linear and angular need to be tuned, and if so start by tuning linear
+		if !(linear.P == 0.0 && linear.I == 0.0 && linear.D == 0.0 &&
+			angular.P == 0.0 && angular.I == 0.0 && angular.D == 0.0) {
+			sb.controlLoopConfig = sb.createControlLoopConfig(linear, angular)
+			if err := sb.setupControlLoops(); err != nil {
+				sb.mu.Lock()
```
Comment (susmitaSanyal) [inline components/base/sensorcontrolled/sensorcontrolled.go:191]: What's the point of unlocking at line 185 and then locking at line 191? Why can't we rely on the mutex lock and unlock from line 124-125?

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 4]
```diff
@@ -158,11 +171,47 @@ func (sb *sensorBase) Reconfigure(ctx context.Context, deps resource.Dependencie
...
+			}
+			// to tune angular PID values, linear PI values must be non-zero
+			fakeConf.Type = typeLinVel
+			sb.logger.Info("tuning angular PID")
+			if err := sb.autoTuningProcess(ctx, fakeConf, angular); err != nil {
+				sb.mu.Lock()
+				return err
+			}
+		}
+		sb.mu.Lock()
```
Comment (susmitaSanyal) [inline components/base/sensorcontrolled/sensorcontrolled.go:214]: This lock feels irrelevant.

**Review by susmitaSanyal — CHANGES_REQUESTED** [review]

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 5]
```diff
@@ -212,7 +261,21 @@ func (sb *sensorBase) Stop(ctx context.Context, extra map[string]interface{}) er
...
+	if err != nil {
+		return false, err
+	}
+	ang, err := sb.velocities.AngularVelocity(ctx, nil)
+	if err != nil {
+		return false, err
+	}
+	if lin.Y == 0.0 && ang.Z == 0.0 {
+		return false, nil
+	}
```
Comment (randhid) [inline components/base/sensorcontrolled/sensorcontrolled.go:277]: There is a case in this driver where an orientation sensor can be provided but not a velocities sensor. We are ignoring that. Also this would mean we're not returning IsMoving correctly from the underlying wheeled base. I'll look into it.

File: control/control_loop.go [hunk 6]
```diff
@@ -306,12 +303,32 @@ func (l *Loop) GetConfig(ctx context.Context) Config {
...
-	return l.tuning
+// MonitorTuning waits for tuning to start, and then returns once it's done.
+func (l *Loop) MonitorTuning(ctx context.Context) error {
+	// wait until tuning has started
+	for {
+		tuning := l.GetTuning(ctx)
+		if tuning {
+			break
+		}
+	}
```
Comment (randhid) [inline control/control_loop.go:314]: this is totally async from the constructor, correct?

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 1 above)
Comment (martha-johnston) [reply components/base/sensorcontrolled/sensorcontrolled.go:180]: hm I think instead I might add a warning that the types aren't correct in an `else` block. since these are coming from the config I don't just want to assume anything not linear is angular

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 2 above)
Comment (martha-johnston) [reply components/base/sensorcontrolled/sensorcontrolled.go:185]: since the tuning process needs to run the motors, this lock prevents the tuning process from starting, so it needs to unlock before anything to do with the tuning, and then relock at the end of the function since there is a defer unlock fro the top of this function. kind of weird I'll add some comments

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 5 above)
Comment (martha-johnston) [reply components/base/sensorcontrolled/sensorcontrolled.go:277]: `IsMoving` for wheeled_base just returns `IsPowered` for all the motors and ignores the power percent. That doesn't work when using the control loop since it holds position, but maybe a better change is to fix `IsMoving` in wheeled base. I originally edited this because I was using `IsMoving` to determine when the tuning is done but I have a different solution now so I can worry about this in a different ticket if you'd prefer. but I also don't mind just changing it now. 

as for using only velocities == nil, that is the condition we use to check if we should start the loop or not on line 174 of this file. and since we are checking the linear and angular velocity we know that `velocities` has to support that. in the case that orientation is being used it will just call `IsMoving` of wheeled base. I _think_ this change is only relevant for using the control loop because is will never not be powered until you manually stop it, so you need to instead know if the velocities are 0 to know if it's moving or not

Comment (martha-johnston) [reply components/base/sensorcontrolled/sensorcontrolled.go:277]: but I haven't looked too closely at the situation in which you only have an orientation reporting sensor so I can look into that closer to understand how that would work with this change

File: control/control_loop.go (hunk 6 above)
Comment (martha-johnston) [reply control/control_loop.go:314]: no, reconfigure doesn't return until the tuning process is done. I'll try to change that right now, good catch

Comment (martha-johnston) [reply control/control_loop.go:314]: fixed

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 4 above)
Comment (susmitaSanyal) [reply components/base/sensorcontrolled/sensorcontrolled.go:207]: Thanks for adding the comments ❤️ This makes reading this function so much easier!!

**Review by susmitaSanyal — APPROVED** [review]
LGTM! approving with the assumption that it passed hardware testing.

File: components/base/sensorcontrolled/sensorcontrolled.go (hunk 5 above)
Comment (randhid) [reply components/base/sensorcontrolled/sensorcontrolled.go:270]: in the constructor for this driver, we also check for orientation reporting sensors and use only that in `spin.go`, so this change ignores that version of the sensor-controlled base completely. Since wheeled odometry is orientation reporting we can test what that's doing to he spin loop as well.

**Review by randhid — CHANGES_REQUESTED** [review]
We should test the IsMoving changes with a orientation reporting sensor, right now you won't return a true is moving call if the constructor has an orientation reporting sensor only.

File: control/control_loop.go [hunk 7]
```diff
@@ -39,7 +39,7 @@ type Loop struct {
 	cancelCtx               context.Context
 	cancel                  context.CancelFunc
 	running                 bool
-	tuning                  bool
```
[the remark this answers is not in the document]
Comment (martha-johnston) [reply control/control_loop.go:42]: Leaving it as is in this ticket if that's alright since I'm already working on a refactor in a different ticket.

File: components/base/wheeled/wheeled_base.go [hunk 8]
```diff
@@ -483,11 +483,14 @@ func (wb *wheeledBase) Stop(ctx context.Context, extra map[string]interface{}) e
...
-		isMoving, _, err := m.IsPowered(ctx, nil)
+		isPowered, pwrPct, err := m.IsPowered(ctx, nil)
 		if err != nil {
 			return false, err
 		}
-		if isMoving {
+		if isPowered {
+			if math.Abs(pwrPct) < 0.01 {
+				return false, err
+			}
```
Comment (martha-johnston) [inline components/base/wheeled/wheeled_base.go:493]: @randhid I reverted sensor base IsMoving to just call real base IsMoving, and added this to wheeled base so it no longer depends on any sensors

Comment (randhid) [reply components/base/wheeled/wheeled_base.go:493]: cool, lets test this out together!

File: components/base/sensorcontrolled/spin.go [hunk 9]
```diff
@@ -46,11 +46,26 @@ func (sb *sensorBase) Spin(ctx context.Context, angleDeg, degsPerSec float64, ex
...
 	}
+
+	if err := sb.opMgr.WaitForSuccess(
+		ctx,
+		yawPollTime,
+		baseMoving,
+	); err != nil {
+		return err
+	}
+
```
Comment (martha-johnston) [inline components/base/sensorcontrolled/spin.go:68]: added some testing logs and this was the behavior I saw, so I do think the issue is that it was immediately returning before the motors had time to start moving
```
sensorcontrolled/spin.go:52	moving = false
sensorcontrolled/spin.go:52	moving = false
sensorcontrolled/spin.go:52	moving = true
sensorcontrolled/spin.go:59	moving = true
sensorcontrolled/spin.go:59	moving = true
```

**Review by randhid — APPROVED** [review]
Can I see the behaviours one last time, and see the state of the skipped test in the controls package? Otherwise lgtm!

**Comment by martha-johnston** [comment]
> Can I see the behaviours one last time, and see the state of the skipped test in the controls package? Otherwise lgtm!

yeah I can show you the behavior whenever you're ready

File: components/base/sensorcontrolled/sensorcontrolled.go [hunk 10]
```diff
@@ -187,6 +229,10 @@ func (sb *sensorBase) isPolling() bool {
 func (sb *sensorBase) MoveStraight(
 	ctx context.Context, distanceMm int, mmPerSec float64, extra map[string]interface{},
 ) error {
+	if sb.loop != nil {
+		sb.loop.Stop()
+		sb.loop = nil
+	}
```
Comment (martha-johnston) [inline components/base/sensorcontrolled/sensorcontrolled.go:235]: @randhid I think this did it, just spammed it with many different calls for a while and haven't seen the errors again

File: components/base/sensorcontrolled/velocities.go [hunk 11]
```diff
@@ -71,15 +68,61 @@ func (sb *sensorBase) updateControlConfig(
...
+
 func (sb *sensorBase) SetVelocity(
 	ctx context.Context, linear, angular r3.Vector, extra map[string]interface{},
 ) error {
 	sb.opMgr.CancelRunning(ctx)
-	// check if a sensor context has been started
-	if sb.sensorLoopDone != nil {
-		sb.sensorLoopDone()
+	// stop any previous base movement
+	if err := sb.Stop(ctx, nil); err != nil {
```
Comment (randhid) [inline components/base/sensorcontrolled/velocities.go:123]: if it is bad when we test with motion's multiple set velocity calls, we should package the stoploop functions and done functions in a helper to cancel the threads without calling stop on the underlying base. Add a comment right above.

**Review by randhid — APPROVED** [review]
